internal/tfdiags/compare.go

```diff
@@ -14,21 +14,91 @@ import "github.com/google/go-cmp/cmp"
 //
 // Example usage:
 //
-//	cmp.Diff(diag1, diag2, tfdiags.DiagnosticComparer())
-var DiagnosticComparer cmp.Option = cmp.Comparer(diagnosticComparer)
+//	cmp.Diff(diag1, diag2, tfdiags.DiagnosticComparer)
+var DiagnosticComparer cmp.Option = cmp.Comparer(diagnosticComparerSimple)
 
-func diagnosticComparer(l, r Diagnostic) bool {
+// diagnosticComparerSimple returns false when a difference is identified between
+// the two Diagnostic arguments.
+func diagnosticComparerSimple(l, r Diagnostic) bool {
 	if l.Severity() != r.Severity() {
 		return false
 	}
 	if l.Description() != r.Description() {
 		return false
 	}
 
+	// Do the diagnostics originate from the same attribute name, if any?
 	lp := GetAttribute(l)
 	rp := GetAttribute(r)
 	if len(lp) != len(rp) {
 		return false
 	}
 	return lp.Equals(rp)
 }
+
+// DiagnosticComparerWithSource returns a cmp.Option that can be used with
+// the package github.com/google/go-cmp/cmp.
+//
+// The comparer expands on the DiagnosticComparer by additionally
+// comparing based on:
+// 1) Source.Subject
+// 2) Source.Context
+//
+// Example usage:
+//
+//	cmp.Diff(diag1, diag2, tfdiags.DiagnosticComparerWithSource)
+var DiagnosticComparerWithSource cmp.Option = cmp.Comparer(diagnosticComparerWithSource)
```

Comment:
Question for reviewer - which tests would need/want to assert about the source data in a diagnostic?

Reply:
I'd like to know too. Otherwise, I wonder if we should wait for such use case before building such a comparer

Reply:
Pretty much anything in the `configs` and `stackconfig` packages should care about the source data, as it is important in those cases that the errors point to the part of the configuration that is invalid. An example of this would be the tests here: https://github.com/hashicorp/terraform/blob/main/internal/configs/config_test.go#L542-L582. But, there should be lots of examples in those packages.

Reply:
It looks like the tests in `configs` use hcl.Diagnostic/hcl.Diagnostics types and in `stackconfig` the diags comparison uses sourceless diags in assertions. So overall it looks like this comparer isn't needed. I'll slim down the PR to just the edits to the existing comparer and then open up the stacked PR for review too.